File: java/04_explain_how_to_use_the_tool/agent-application/src/main/java/com/workshop/agent/application/Agent.java
package com.workshop.agent.application;

import java.util.List;

public class Agent {
    private final UserInput input;
    private final Display display;
    private final LanguageModel model;

    public Agent(UserInput input, LanguageModel model, Display display) {
        this.input = input;
        this.display = display;
        this.model = model;
    }

    public void run() {
        String userInput = input.getInput();
        Message userMessage = new Message("user", userInput);
        display.show(userMessage);
        Message answer = model.prompt(List.of(userMessage));
        display.show(answer);
    }
}

File: csharp/template/Agent/Agent.Application/Agent.cs
namespace Agent.Application;

public class Agent()
{
    private readonly string _systemPrompt;
    private readonly IUserInput _input;
    private readonly IDisplay _display;
    private readonly ILanguageModel _model;
    private readonly ITool _tool;

    public Agent(string systemPrompt, IUserInput input, ILanguageModel model, ITool tool, IDisplay display) : this()
    {
        _systemPrompt = systemPrompt;
        _input = input;
        _display = display;
        _model = model;
        _tool = tool;
    }

    public void Run()
    {
        var context = new List<Message>
        {
            new("system", _systemPrompt)
        };

        while (true)
        {
            var userInput = _input.GetInput();
            if (string.IsNullOrWhiteSpace(userInput))
            {
                break;
            }

            var userMessage = new Message("user", userInput);
            _display.Show(userMessage);
            context.Add(userMessage);

            var answer = _model.Prompt(context);
            _display.Show(answer);
            context.Add(answer);

            while (true)
            {
                var toolResult = _tool.ParseAndExecute(answer.Content);
                if (toolResult == null)
                {
                    break;
                }

                var toolResultMessage = new Message("user", toolResult);
                context.Add(toolResultMessage);
                _display.Show(toolResultMessage);
                answer = _model.Prompt(context);
                _display.Show(answer);
                context.Add(answer);
            }
        }
    }
}